include/zephyr/sys/async.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Public APIs for Async utilities
+ *
+ * Functions and datatypes useful for safely communicating to a thread or work
+ * queue of an asynchronous functions completion.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ASYNC_H_
+#define ZEPHYR_INCLUDE_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**
+ * @brief Async operation callback type
+ *
+ * Common function pointer type used for asynchronous completion callbacks
+ * and implemented for some commonly used IPC notification mechanisms.
+ *
+ * May also be implemented by an application with the understanding
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
+
+#ifdef CONFIG_ASYNC
```

Comment:
Should the API really be conditional? The implementation, I can understand (even if it wont result in anything but compile-time overhead if unconditional) - but the API?